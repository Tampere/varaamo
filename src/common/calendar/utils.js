import moment from 'moment';
import get from 'lodash/get';

/**
 * getDefaultSelectedTimeRange()
 * Get default selected time range from reservation data.
 * @returns {Object} selected time range
 */
export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end,
  } : null;
};


/**
 * isTimeRangeOverMaxPeriod();
 * Reservation cannot be longer than the resources max period if max period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than max_period or not
 */
export const isTimeRangeOverMaxPeriod = (resource, start, end, isStaff = false) => {
  const period = get(resource, 'max_period', null);

  if (!isStaff && period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') > durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * isTimeRangeUnderMinPeriod();
 * Reservation must longer than the resources min period if min period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than min_period or not
 */
export const isTimeRangeUnderMinPeriod = (resource, start, end) => {
  const period = get(resource, 'min_period', null);

  if (period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') < durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * addPeriodToTime()
 * Add period to input time
 * For example: period 00:30:00, time: moment(12:00:00) -> moment(12:30:00)
 *
 * @param {Moment} time
 * @param {string} period
 * @returns {Moment}
 */

const addPeriodToTime = (time, period) => {
  if (period) {
    const periodInMinute = moment.duration(period).as('minutes');
    return time.add(periodInMinute, 'minutes');
  }

  // Return original if no period
  return time;
};

/**
 * getMinPeriodTimeRange
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMinPeriodTimeRange = (resource, start, end) => {
  const minPeriod = get(resource, 'min_period', null);

  if (!minPeriod) {
    return {
      start, end,
    };
  }

  const minPeriodEnd = addPeriodToTime(moment(start), minPeriod);

  if (minPeriodEnd.diff(moment(end)) > 0) {
    // time range smaller than min_period, return min_period
    return {
      start,
      end: minPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};

/**
 * getMaxPeriodEndTime
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMaxPeriodTimeRange = (resource, start, end, isStaff) => {
  const maxPeriod = get(resource, 'max_period', null);

  if (!maxPeriod || isStaff) {
    return {
      start, end,
    };
  }

  const maxPeriodEnd = addPeriodToTime(moment(start), maxPeriod);

  if (moment(end).diff(maxPeriodEnd) > 0) {
    // time range is over max_period not allowed, return max_period
    return {
      start,
      end: maxPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};
