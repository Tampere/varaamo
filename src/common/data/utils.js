import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';
import camelCase from 'lodash/camelCase';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};

/**
 * Trim empty values from data object
 *
 * @param {Object} data
 * @returns {Object} Trimmed empty values object
 */

export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};

export const toCamelCase = (obj) => {
  if (!obj) return {};

  const camelCasedObj = {};

  Object.keys(obj).forEach((key) => {
    const camelCasedKey = camelCase(key);
    camelCasedObj[camelCasedKey] = obj[key];
  });

  return camelCasedObj;
};
