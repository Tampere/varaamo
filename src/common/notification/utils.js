import { NotificationManager } from 'react-notifications';

import { NOTIFICATION_TYPE } from './constants';
/**
 * createNotification()
 * Creating notification popups
 *
 * @param {string} type
 * @param {string} [message='']
 * @param {string} [title='']
 * @param {number} [fadeTime=2000]
 * @param {function} callback
 * @returns {void}
 */
export const createNotification = (type, message = '', title = '', fadeTime = 2000, callback) => {
  switch (type) {
    case NOTIFICATION_TYPE.INFO:
      NotificationManager.info(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.SUCCESS:
      NotificationManager.success(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.WARNING:
      NotificationManager.warning(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.ERROR:
      NotificationManager.error(message, title, fadeTime, callback);
      break;
    default:
      NotificationManager.info(message, title, fadeTime, callback);
  }
};
