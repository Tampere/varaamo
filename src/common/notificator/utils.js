import moment from 'moment';

/**
 * createSelectOptions()
 * Create option arrays for selects.
 * @returns {Object} with select option arrays
 */
export const createSelectOptions = () => {
  const targetOptions = [
    { value: 'staff', label: 'Staff' },
    { value: 'user', label: 'User' },
    { value: 'all', label: 'All' },
  ];
  const urgencyOptions = [
    { value: 'common', label: 'Common' },
    { value: 'warning', label: 'Warning' },
    { value: 'danger', label: 'Danger' },
  ];
  const activeOptions = [
    { value: true, label: 'True' },
    { value: false, label: 'False' },
  ];

  return {
    targetOptions,
    urgencyOptions,
    activeOptions,
  };
};


/**
 * formatValuesForUse()
 * Format notification values to usable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForUse = (notification) => {
  const { targetOptions, urgencyOptions, activeOptions } = createSelectOptions();
  const formattedNotification = { ...notification };
  formattedNotification.target = targetOptions.find(option => option.value === notification.target);
  formattedNotification.urgency = urgencyOptions.find(option => option.value === notification.urgency);
  formattedNotification.active = activeOptions.find(option => option.value === notification.active);
  formattedNotification.until = moment(notification.until).toDate();
  return formattedNotification;
};

/**
 * formatValuesForDatabase()
 * Format notification values to savable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForDatabase = (notification) => {
  const formattedNotification = { ...notification };
  formattedNotification.target = notification.target.value;
  formattedNotification.urgency = notification.urgency.value;
  formattedNotification.active = notification.active.value;
  formattedNotification.until = moment(notification.until).format('YYYYMMDDTHHmmss');
  return formattedNotification;
};
