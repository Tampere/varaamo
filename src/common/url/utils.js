/**
 * Builds search string from an object.
 * @param search {object}
 * @returns {string}
 */
export const getSearch = (search) => {
  return new URLSearchParams(search).toString();
};

/**
 * Getter for link string with given path and query params.
 * @param path {string}
 * @param search {object}
 * @returns {string}
 */
export const getLinkString = (path, search = null) => {
  let searchString = '';

  if (search) {
    searchString = getSearch(search);
  }

  return `${path}${search ? `?${searchString}` : ''}`;
};
