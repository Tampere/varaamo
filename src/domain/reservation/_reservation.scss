@import 'modal/reservationInformationModal';

@import 'manage/page/manageReservationsPage';
@import 'manage/filters/manageReservationsFilters';
@import 'manage/list/manageReservationsList';
@import 'manage/status/manageReservationsStatus';
