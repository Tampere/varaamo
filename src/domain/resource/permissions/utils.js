import difference from 'lodash/difference';
import get from 'lodash/get';

import {
  resourceRoles, resourcePermissionTypes, resourcePermissionsByRole, UI_UNIT_STAFF_ROLES,
} from './constants';

const permissionOptions = Object.values(resourcePermissionTypes);
const allValuesInArray = (subset, superset) => difference(subset, superset).length === 0;

export function hasPermissionForResource(resourceRole, requiredPermissions) {
  // undefined roles don't receive any permissions implicitly
  if (!resourceRole) {
    return false;
  }

  if (!Object.values(resourceRoles).includes(resourceRole)) {
    throw Error(`resourceRole needs to be one of: ${Object.values(resourceRoles).join(', ')}.`);
  }

  if (!Array.isArray(requiredPermissions) && typeof requiredPermissions !== 'string') {
    throw Error('requiredPermissions has to be either a string or an array.');
  }

  if (
    Array.isArray(requiredPermissions)
    && !allValuesInArray(requiredPermissions, permissionOptions)
  ) {
    const unknownPermissions = difference(requiredPermissions, permissionOptions);

    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${unknownPermissions.join(', ')}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  if (typeof requiredPermissions === 'string' && !permissionOptions.includes(requiredPermissions)) {
    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${requiredPermissions}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  const rolePermissions = resourcePermissionsByRole[resourceRole];

  if (Array.isArray(requiredPermissions)) {
    return allValuesInArray(requiredPermissions, rolePermissions);
  }

  return rolePermissions.includes(requiredPermissions);
}

const getUserPermissions = resource => (
  get(resource, 'user_permissions', false)
  || get(resource, 'userPermissions', undefined)
);

// Returns the role with the most permissions or null.
export function getUnitRoleFromResource(resource) {
  const userPermissions = getUserPermissions(resource);

  if (!userPermissions) {
    return null;
  }

  if (userPermissions.is_admin || userPermissions.isAdmin) {
    return resourceRoles.UNIT_ADMINISTRATOR;
  }

  if (userPermissions.is_manager || userPermissions.isManager) {
    return resourceRoles.UNIT_MANAGER;
  }

  if (userPermissions.is_viewer || userPermissions.isViewer) {
    return resourceRoles.UNIT_VIEWER;
  }

  return null;
}

export function getIsUnitStaff(unitRole) {
  return UI_UNIT_STAFF_ROLES.includes(unitRole);
}
