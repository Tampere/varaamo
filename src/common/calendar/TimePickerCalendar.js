import React, { Component } from 'react';
import FullCalendar from '@fullcalendar/react';
import PropTypes from 'prop-types';
import timeGridPlugin from '@fullcalendar/timegrid';
import momentTimezonePlugin from '@fullcalendar/moment-timezone';
import interactionPlugin from '@fullcalendar/interaction';
import enLocale from '@fullcalendar/core/locales/en-gb';
import svLocale from '@fullcalendar/core/locales/sv';
import fiLocale from '@fullcalendar/core/locales/fi';
import moment from 'moment';
import get from 'lodash/get';
import classNames from 'classnames';
import isEmpty from 'lodash/isEmpty';

import CalendarLegend from './legend/CalendarLegend';
import * as resourceUtils from '../../domain/resource/utils';
import constants from '../../../app/constants/AppConstants';
import injectT from '../../../app/i18n/injectT';
import * as calendarUtils from './utils';
import { createNotification } from '../notification/utils';
import { NOTIFICATION_TYPE } from '../notification/constants';

const NEW_RESERVATION = 'NEW_RESERVATION';

moment.tz.setDefault(constants.TIME_ZONE);
// Re-apply moment-timezone default timezone, cause FullCalendar import have override the import

class TimePickerCalendar extends Component {
  calendarRef = React.createRef();

  static propTypes = {
    date: PropTypes.string,
    disabled: PropTypes.bool,
    isStaff: PropTypes.bool.isRequired,
    resource: PropTypes.object.isRequired,
    onDateChange: PropTypes.func.isRequired,
    t: PropTypes.func.isRequired,
    onTimeChange: PropTypes.func.isRequired,
    locale: PropTypes.string.isRequired,
    edittingReservation: PropTypes.object,
  };

  state = {
    viewType: 'timeGridWeek',
    selected: calendarUtils.getDefaultSelectedTimeRange(this.props.edittingReservation),
    header: {
      left: 'myPrev,myNext,myToday',
      center: 'title',
      right: 'timeGridDay,timeGridWeek',
    },
  };

  componentDidUpdate(prevProps, prevState) {
    const { date } = this.props;
    const { viewType } = this.state;

    const calendarApi = this.calendarRef.current.getApi();

    if (date !== prevProps.date) {
      calendarApi.gotoDate(date);
    }
    if (viewType !== prevState.viewType) {
      calendarApi.changeView(viewType);
    }
  }

  isSelectionValid = (selection) => {
    const { resource } = this.props;

    const events = this.getReservedEvents();
    return resourceUtils.isTimeRangeReservable(resource, selection.start, selection.end, events);
  };

  onChange = (selected) => {
    this.setState({ selected });

    // Invoke select handler callback from props
    this.props.onTimeChange(selected);
  }

  onCancel = () => {
    // Revert to default timerange if cancel
    const defaultSelectedTimeRange = calendarUtils.getDefaultSelectedTimeRange(this.props.edittingReservation);
    this.onChange(defaultSelectedTimeRange);
  }

  onEventRender = (info) => {
    // add cancel button for new selected event
    let duration;

    if (info.event.id === NEW_RESERVATION) {
      const cancelBtn = document.createElement('span');
      cancelBtn.classList.add('app-TimePickerCalendar__cancelEvent');
      cancelBtn.addEventListener('click', () => this.onCancel(), { once: true });
      info.el.append(cancelBtn);
      duration = this.getDurationText(info.event);
    } else if (info.event.id === '') {
      duration = this.getDurationText(info.event);
    }

    if (duration) {
      const eventDuration = document.createElement('span');
      eventDuration.textContent = duration;
      eventDuration.classList.add('app-TimePickerCalendar__maxDuration');
      info.el.append(eventDuration);
    }
  };

  onSelect = (selectionInfo) => {
    const { t } = this.props;
    const calendarApi = this.calendarRef.current.getApi();
    calendarApi.unselect();
    // Clear FullCalendar select tooltip

    const selectable = this.getSelectableTimeRange(selectionInfo);
    // Make sure selected time alway furfill period check.

    const isValid = this.isSelectionValid(selectable);

    if (isValid) {
      this.onChange(selectable);
    } else {
      // Display error notifications if selection is not valid
      createNotification(NOTIFICATION_TYPE.ERROR, t('Notifications.selectTimeToReserve.warning'));
    }
  };

  // Check if event resize allowed
  onEventResize = (selectionInfo) => {
    const { event } = selectionInfo;
    const selectable = this.getSelectableTimeRange(event, selectionInfo);
    this.onChange(selectable);
  }

  onDatesRender = (info) => {
    const { viewType, header } = this.state;
    let view = info.view.type;
    let headerConfig = header;

    // 768
    if (window.innerWidth < 768) {
      // mobile view config
      view = 'timeGridDay';
      headerConfig = {
        left: 'myPrev,myNext,myToday',
        center: 'title',
        right: 'timeGridDay,timeGridWeek',
      };
    }
    if (viewType !== view) {
      this.setState({ viewType: view, header: headerConfig });
    }
  }

  /**
   * To ensure the selected time range will always between min and max period
   * Return a selectable timerange if input timerange is somehow not between min and max period
   * Display info text to user about those changes.
   *
   * @return  {Object}  Normalized selected time range
   */
  getSelectableTimeRange = (selected, eventCallback) => {
    const { resource, isStaff, t } = this.props;

    let selectable = {
      start: selected.start,
      end: selected.end,
    };

    const isUnderMinPeriod = calendarUtils.isTimeRangeUnderMinPeriod(
      resource, selectable.start, selectable.end,
    );

    const isOverMaxPeriod = calendarUtils.isTimeRangeOverMaxPeriod(
      resource, selectable.start, selectable.end, isStaff,
    );

    if (isUnderMinPeriod) {
      const minPeriod = get(resource, 'min_period', null);
      const minPeriodDuration = moment.duration(minPeriod).asHours();

      if (eventCallback) {
        eventCallback.revert();
        createNotification(
          NOTIFICATION_TYPE.INFO, t('TimePickerCalendar.info.minPeriodText', { duration: minPeriodDuration }),
        );
      }

      selectable = calendarUtils.getMinPeriodTimeRange(resource, selected.start, selected.end);
      // Make sure selected time will always bigger than min period
    }

    if (isOverMaxPeriod) {
      const maxPeriod = get(resource, 'max_period', null);
      const maxPeriodDuration = moment.duration(maxPeriod).asHours();

      if (eventCallback) {
        eventCallback.revert();
      }

      createNotification(
        NOTIFICATION_TYPE.INFO, t('TimePickerCalendar.info.maxPeriodText', { duration: maxPeriodDuration }),
      );

      selectable = calendarUtils.getMaxPeriodTimeRange(
        resource, selectable.start, selectable.end, isStaff,
      );
      // Make sure selected time will always smaller than max period
    }

    return selectable;
  }

  getDurationText = (selected) => {
    const { resource } = this.props;
    const start = moment(selected.start);
    const end = moment(selected.end);
    const duration = moment.duration(end.diff(start));

    let maxDurationText = '';

    if (resource.max_period) {
      const maxDuration = get(resource, 'max_period', null);
      const maxDurationSeconds = moment.duration(maxDuration).asSeconds();
      maxDurationText = `(${maxDurationSeconds / 3600}h max)`;
    }

    return `${duration / 3600000}h ${maxDurationText}`;
  };

  getSelectedDateText = () => {
    const { t, resource } = this.props;
    const { selected } = this.state;

    if (selected) {
      const start = moment(selected.start);
      const end = moment(selected.end);
      const price = resourceUtils.getReservationPrice(selected.start, selected.end, resource);

      const tVariables = {
        date: start.format('dd D.M.Y'),
        start: start.format('HH:mm'),
        end: end.format('HH:mm'),
        duration: this.getDurationText(),
        price,
      };

      if (price) {
        return t('TimePickerCalendar.selectedDateValueWithPrice', tVariables);
      }

      return t('TimePickerCalendar.selectedDateValue', tVariables);
    }

    return '';
  };

  getReservedEvents = () => {
    const { resource, date } = this.props;

    const getClassNames = (reservation) => {
      const isOwn = get(reservation, 'is_own', false);
      return classNames('app-TimePickerCalendar__event', {
        'app-TimePickerCalendar__event--reserved': !isOwn,
      });
    };
    // Add the resources reservations as normal FullCalendar events.
    const reservations = get(resource, 'reservations', []);
    const events = isEmpty(reservations) ? [] : reservations.map(reservation => ({
      classNames: [getClassNames(reservation)],
      editable: false,
      id: reservation.id,
      start: moment(reservation.begin).toDate(),
      end: moment(reservation.end).toDate(),
    }));

    // Check resources reservation rules and disable days if needed.
    const now = moment();
    const momentDate = moment(date)
      .startOf('week');

    for (let i = 0; i < 7; i++) {
      if (
        momentDate.isBefore(now, 'date')
        || !resourceUtils.isDateReservable(resource, momentDate.format(constants.DATE_FORMAT))
      ) {
        events.push({
          allDay: true,
          classNames: [
            'app-TimePickerCalendar__backgroundEvent',
            'app-TimePickerCalendar__restrictedDate',
          ],
          id: momentDate.format(constants.DATE_FORMAT),
          start: momentDate.toDate(),
          end: momentDate.toDate(),
          rendering: 'background',
        });
      }
      momentDate.add(1, 'day');
    }
    return events;
  }

  getCalendarOptions = () => {
    const isDisabled = this.props.disabled;

    return {
      timeZone: constants.TIME_ZONE,
      height: 'auto',
      editable: !isDisabled,
      eventConstraint: 'businessHours',
      eventOverlap: false,
      firstDay: 1,
      locale: this.props.locale,
      locales: [enLocale, svLocale, fiLocale],
      nowIndicator: true,
      plugins: [timeGridPlugin, momentTimezonePlugin, interactionPlugin],
      selectable: !isDisabled,
      selectOverlap: false,
      selectConstraint: 'businessHours',
      selectMirror: true,
      dragScroll: true,
      slotLabelFormat: {
        hour: 'numeric',
        minute: '2-digit',
        omitZeroMinute: false,
        meridiem: 'short',
      },
      unselectAuto: false,
      longPressDelay: 250,
      eventLongPressDelay: 20,
      selectLongPressDelay: 200,
      // Almost invoke click event on mobile immediatelly without any delay
    };
  };

  getEvents = () => {
    const {
      resource, isStaff,
    } = this.props;
    const { selected } = this.state;

    const events = this.getReservedEvents();
    if (selected) {
      const webEventSelected = window.innerWidth > 768 ? 'fc-selected' : '';
      events.push({
        classNames: [
          'app-TimePickerCalendar__event',
          'app-TimePickerCalendar__newReservation',
          webEventSelected,
        ],
        editable: true,
        durationEditable: !calendarUtils.isTimeRangeOverMaxPeriod(
          resource, selected.start, selected.end, isStaff,
        ),
        id: NEW_RESERVATION,
        ...selected,
      });
    }
    return events;
  };

  render() {
    const {
      date,
      onDateChange,
      resource,
      t,
      disabled,
    } = this.props;
    const { viewType, header } = this.state;
    const addValue = viewType === 'timeGridWeek' ? 'w' : 'd';
    const slotTallness = resourceUtils.getSlotTallness(resource);

    return (
      <div className={classNames(['app-TimePickerCalendar', {
        'app-TimePickerCalendar--disabled': disabled,
        [`app-TimePickerCalendar--${slotTallness}-slots`]: slotTallness,
      }])}
      >
        <FullCalendar
          {...this.getCalendarOptions()}
          allDaySlot={false}
          businessHours={resourceUtils.getFullCalendarBusinessHours(resource, date)}
          customButtons={{
            myPrev: {
              text: ' ',
              click: () => onDateChange(moment(date).subtract(1, addValue).toDate()),
            },
            myNext: {
              text: ' ',
              click: () => onDateChange(moment(date).add(1, addValue).toDate()),
            },
            myToday: {
              text: t('TimePickerCalendar.info.today'),
              click: () => onDateChange(this.calendarRef.current.calendar.view.initialNowDate),
            },
          }}
          datesRender={this.onDatesRender}
          defaultDate={date}
          eventDrop={this.onEventResize}
          eventRender={this.onEventRender}
          eventResize={this.onEventResize}
          events={this.getEvents()}
          header={header}
          maxTime={resourceUtils.getFullCalendarMaxTime(resource, date, viewType)}
          minTime={resourceUtils.getFullCalendarMinTime(resource, date, viewType)}
          onDatesRender={this.onDatesRender}
          ref={this.calendarRef}
          select={this.onSelect}
          slotDuration={resourceUtils.getFullCalendarSlotDuration(resource, date, viewType)}
          slotLabelInterval={resourceUtils.getFullCalendarSlotLabelInterval(resource)}
        />
        <CalendarLegend />
        {disabled && (
          <div className="app-TimePickerCalendar__blocker">
            <div className="app-TimePickerCalendar__inline-modal">{t('TimePickerCalendar.info.displayedMessage')}</div>
          </div>
        )}
      </div>
    );
  }
}

export default injectT(TimePickerCalendar);
