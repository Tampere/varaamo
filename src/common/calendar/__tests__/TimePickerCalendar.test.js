import React from 'react';
import toJson from 'enzyme-to-json';

import TimePickerCalendar from '../TimePickerCalendar';
import resource from '../../data/fixtures/resource';
import reservation from '../../data/fixtures/reservation';
import { globalDateMock, shallowWithIntl } from '../../../../app/utils/testUtils';

describe('TimePickerCalendar', () => {
  globalDateMock();

  const defaultProps = {
    resource: resource.build(),
    date: '2019-08-15',
    isStaff: false,
    onDateChange: jest.fn(),
    onReserve: jest.fn(),
    onTimeChange: jest.fn(),
  };

  const selected = {
    start: '2019-09-05T10:30:00+03:00',
    end: '2019-09-05T12:00:00+03:00',
  };


  const getWrapper = props => shallowWithIntl(<TimePickerCalendar {...defaultProps} {...props} />);

  describe('FullCalendar', () => {
    test('render normally', () => {
      const wrapper = getWrapper();

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('edittingReservation', () => {
    const defaultSelected = {
      begin: '2019-09-05T10:30:00+03:00',
      end: '2019-09-05T12:00:00+03:00',
    };

    const edittingReservation = reservation.build(defaultSelected);

    const wrapper = getWrapper({ edittingReservation });

    test('will have selected state intitialize from edittingReservation props', () => {
      expect(wrapper.state('selected')).toEqual(selected);
    });

    test('will populate new reservation event with timeslot selected from edittingReservation', () => {
      const events = wrapper.instance().getEvents();
      expect(events[0].classNames[1]).toContain('newReservation');
    });

    test('will render edittingReservation slot when cancel current selected slot', () => {
      wrapper.setState({
        selected: {
          start: '2019-09-05T13:30:00+03:00',
          end: '2019-09-05T14:00:00+03:00',
        },
      });

      // select random time range
      const instance = wrapper.instance();

      // override calendar context with mock
      instance.calendarRef = {
        current: {
          getApi: () => ({
            unselect: jest.fn(),
          }),
        },
      };

      wrapper.update();

      // Mock on clicking cancel button
      instance.onCancel();

      // Return data match default from reservation
      expect(wrapper.state('selected')).toEqual(selected);
    });
  });
});
