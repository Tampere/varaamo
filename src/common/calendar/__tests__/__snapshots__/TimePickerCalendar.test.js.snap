// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePickerCalendar FullCalendar render normally 1`] = `
<div
  className="app-TimePickerCalendar"
>
  <FullCalendar
    allDaySlot={false}
    businessHours={
      Array [
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
      ]
    }
    customButtons={
      Object {
        "myNext": Object {
          "click": [Function],
          "text": " ",
        },
        "myPrev": Object {
          "click": [Function],
          "text": " ",
        },
        "myToday": Object {
          "click": [Function],
          "text": "TimePickerCalendar.info.today",
        },
      }
    }
    datesRender={[Function]}
    defaultDate="2019-08-15"
    dragScroll={true}
    editable={true}
    eventConstraint="businessHours"
    eventDrop={[Function]}
    eventLongPressDelay={20}
    eventOverlap={false}
    eventRender={[Function]}
    eventResize={[Function]}
    events={Array []}
    firstDay={1}
    header={
      Object {
        "center": "title",
        "left": "myPrev,myNext,myToday",
        "right": "timeGridDay,timeGridWeek",
      }
    }
    height="auto"
    locale="en"
    locales={
      Array [
        Object {
          "code": "en-gb",
          "week": Object {
            "dow": 1,
            "doy": 4,
          },
        },
        Object {
          "allDayText": "Heldag",
          "buttonText": Object {
            "day": "Dag",
            "list": "Program",
            "month": "Månad",
            "next": "Nästa",
            "prev": "Förra",
            "today": "Idag",
            "week": "Vecka",
          },
          "code": "sv",
          "eventLimitText": "till",
          "noEventsMessage": "Inga händelser att visa",
          "week": Object {
            "dow": 1,
            "doy": 4,
          },
          "weekLabel": "v.",
        },
        Object {
          "allDayText": "Koko päivä",
          "buttonText": Object {
            "day": "Päivä",
            "list": "Tapahtumat",
            "month": "Kuukausi",
            "next": "Seuraava",
            "prev": "Edellinen",
            "today": "Tänään",
            "week": "Viikko",
          },
          "code": "fi",
          "eventLimitText": "lisää",
          "noEventsMessage": "Ei näytettäviä tapahtumia",
          "week": Object {
            "dow": 1,
            "doy": 4,
          },
          "weekLabel": "Vk",
        },
      ]
    }
    longPressDelay={250}
    maxTime="17:00:00"
    minTime="07:00:00"
    nowIndicator={true}
    onDatesRender={[Function]}
    plugins={
      Array [
        Object {
          "calendarInteractions": Array [],
          "cmdFormatter": undefined,
          "componentInteractions": Array [],
          "datePointTransforms": Array [],
          "dateSelectionTransformers": Array [],
          "dateSpanTransforms": Array [],
          "defaultView": "timeGridWeek",
          "deps": Array [],
          "elementDraggingImpl": undefined,
          "eventDefMutationAppliers": Array [],
          "eventDefParsers": Array [],
          "eventDragMutationMassagers": Array [],
          "eventDropTransformers": Array [],
          "eventResizeJoinTransforms": Array [],
          "eventSourceDefs": Array [],
          "externalDefTransforms": Array [],
          "id": "6",
          "isDraggableTransformers": Array [],
          "isPropsValid": null,
          "namedTimeZonedImpl": undefined,
          "optionChangeHandlers": Object {},
          "recurringTypes": Array [],
          "reducers": Array [],
          "themeClasses": Object {},
          "viewContainerModifiers": Array [],
          "viewPropsTransformers": Array [],
          "views": Object {
            "timeGrid": Object {
              "allDaySlot": true,
              "class": [Function],
              "slotDuration": "00:30:00",
              "slotEventOverlap": true,
            },
            "timeGridDay": Object {
              "duration": Object {
                "days": 1,
              },
              "type": "timeGrid",
            },
            "timeGridWeek": Object {
              "duration": Object {
                "weeks": 1,
              },
              "type": "timeGrid",
            },
          },
        },
        Object {
          "calendarInteractions": Array [],
          "cmdFormatter": undefined,
          "componentInteractions": Array [],
          "datePointTransforms": Array [],
          "dateSelectionTransformers": Array [],
          "dateSpanTransforms": Array [],
          "defaultView": "",
          "deps": Array [],
          "elementDraggingImpl": undefined,
          "eventDefMutationAppliers": Array [],
          "eventDefParsers": Array [],
          "eventDragMutationMassagers": Array [],
          "eventDropTransformers": Array [],
          "eventResizeJoinTransforms": Array [],
          "eventSourceDefs": Array [],
          "externalDefTransforms": Array [],
          "id": "7",
          "isDraggableTransformers": Array [],
          "isPropsValid": null,
          "namedTimeZonedImpl": [Function],
          "optionChangeHandlers": Object {},
          "recurringTypes": Array [],
          "reducers": Array [],
          "themeClasses": Object {},
          "viewContainerModifiers": Array [],
          "viewPropsTransformers": Array [],
          "views": Object {},
        },
        Object {
          "calendarInteractions": Array [
            [Function],
          ],
          "cmdFormatter": undefined,
          "componentInteractions": Array [
            [Function],
            [Function],
            [Function],
            [Function],
          ],
          "datePointTransforms": Array [],
          "dateSelectionTransformers": Array [],
          "dateSpanTransforms": Array [],
          "defaultView": "",
          "deps": Array [],
          "elementDraggingImpl": [Function],
          "eventDefMutationAppliers": Array [],
          "eventDefParsers": Array [],
          "eventDragMutationMassagers": Array [],
          "eventDropTransformers": Array [],
          "eventResizeJoinTransforms": Array [],
          "eventSourceDefs": Array [],
          "externalDefTransforms": Array [],
          "id": "8",
          "isDraggableTransformers": Array [],
          "isPropsValid": null,
          "namedTimeZonedImpl": undefined,
          "optionChangeHandlers": Object {},
          "recurringTypes": Array [],
          "reducers": Array [],
          "themeClasses": Object {},
          "viewContainerModifiers": Array [],
          "viewPropsTransformers": Array [],
          "views": Object {},
        },
      ]
    }
    select={[Function]}
    selectConstraint="businessHours"
    selectLongPressDelay={200}
    selectMirror={true}
    selectOverlap={false}
    selectable={true}
    slotDuration="00:30:00"
    slotLabelFormat={
      Object {
        "hour": "numeric",
        "meridiem": "short",
        "minute": "2-digit",
        "omitZeroMinute": false,
      }
    }
    slotLabelInterval="01:00:00"
    timeZone="Europe/Helsinki"
    unselectAuto={false}
  />
  <InjectT(CalendarLegend) />
</div>
`;
